### PR TITLE
Introduce a new WiFi manager

### PR DESCRIPTION
**Pull request by kelvincheung**
Introduce a new WiFi manager to support general-purpose full functional
Wi-Fi control plane, including both STA and SoftAP mode.
    
    - Service-orientation, modular design, general purpose
    - Dedicated state machine for each mode,
      which has the knowledge of the Wi-Fi status
    - Support OPEN/WPA-PSK/WPA2-PSK
    - Support RTT
    - Support delete station and blacklist function of SoftAP
    - Support configuration saving
    - Provide blocked APIs to simplify the process of asynchronous actions
    - Support multiple clients
    - POSIX compliant
    - Easy to extend & maintain
    
This manager provides another option for offload Wi-Fi devices.

### CODE REVIEW

**Comment by kelvincheung** [comment]
The new Wi-Fi manager goes with Unisoc UWP566x Wi-Fi driver.(https://github.com/zephyrproject-rtos/zephyr/pull/15014)

**Review by jukkar — CHANGES_REQUESTED** [review]
Most of this stuff is generic and should be added to `subsys/net/l2/wifi/wifi_{mgmt|shell}.c` files.

**Comment by kelvincheung** [comment]
@jukkar 

> Most of this stuff is generic and should be added to `subsys/net/l2/wifi/wifi_{mgmt|shell}.c` files.

Yes, this new WiFi Manager is for general purpose, and can be used by other WiFi vendors.
Meanwhile, it's a dedicated module, and has no relation with existed network management.
Perhaps it's the time to create a new subsystem for WiFi.

File: subsys/wifi/shell/shell.c [hunk 1]
```diff
@@ -0,0 +1,612 @@
...
+		return -EINVAL;
+
+	ret = strtomac(argv[1], mac_addr);
+	if (ret) {
+		printf("invalid MAC address!\n");
+		return ret;
+	}
+
+	if (is_broadcast_ether_addr(mac_addr))
+		printf("Deauth all stations!\n");
```
Comment (jarz-nordic) [inline subsys/wifi/shell/shell.c]: Please avoid using printf in command handlers. You have dedicated shell API for that:
Function: shell_fprintf or macros: `shell_print`, `shell_error`, `shell_info`, `shell_warn`.
Reason for that:
1. Shell is taking care to synchronize all messages on the screen. When you use printk you will compete for (usually) uart and displayed text can be in not correct order.
2. Shell is multiinstance so possibly you will not see this message at all. When you run shell on RTT and printf is connected to UART.

**Review by jakub-uC — DISMISSED** [review]

File: subsys/wifi/shell/shell.c (hunk 1 above)
Comment (kelvincheung) [reply subsys/wifi/shell/shell.c]: @jarz-nordic 
Done, thanks!

File: subsys/wifi/shell/shell.c [hunk 2]
```diff
@@ -0,0 +1,626 @@
...
+
+	memset(&conf, 0, sizeof(conf));
+
+	optind = 0;
+	while ((choice = getopt(argc, argv, "a:b:c:m:n:p:w:")) != -1) {
+		switch (choice) {
+		case 'a':
+			conf.autorun = atoi(optarg);
+			if (!conf.autorun) {
+				shell_print(shell, "invalid autorun!");
```
Comment (jarz-nordic) [inline subsys/wifi/shell/shell.c]: Just a proposal. It would be more neat to use `shell_error` in error messages.

Comment (kelvincheung) [reply subsys/wifi/shell/shell.c]: @jarz-nordic
Done, thanks!

**Review by jakub-uC — COMMENTED** [review]
shell part looks ok.

**Comment by carlescufi** [comment]
> shell part looks ok.

Can you approve or dimissis your request for changes @jarz-nordic ?

**Comment by dleach02** [comment]
Something to think about (piggy backing on @jukkar comment) is whether we want to consider a common WiFi interface. The existing WiFi management assumes the underlying WiFi basically has a complete offload capability of AP/STA modes while this PR appears to implement the STA/AP mode up in the host (which is useful). So should we consider a normalized/common WiFi interface for the application layers to use where offload can be optionally used for devices that support it or use the code here to implement the various modes?

**Review by dleach02 — COMMENTED** [review]
Any plans for supporting the enterprise security modes?

**Comment by kelvincheung** [comment]
> Something to think about (piggy backing on @jukkar comment) is whether we want to consider a common WiFi interface. The existing WiFi management assumes the underlying WiFi basically has a complete offload capability of AP/STA modes while this PR appears to implement the STA/AP mode up in the host (which is useful). So should we consider a normalized/common WiFi interface for the application layers to use where offload can be optionally used for devices that support it or use the code here to implement the various modes?

Yes, the WiFi management does well for the offload WiFi devices.I've ever tried to adapt our WiFi device for it before. However, it has no knowledge of the WiFi status and leaves the synchronization problem to the users. Let's take the auto-connect scenario for example. If you want the WiFi to auto-connect the specified AP once the board is powered on, at least two operations (scan and connect) will be necessary. Using WiFi management, you will need to use a synchronous mechanism to wait for scan done before issuing connect, which is inconvenient obviously. Furthermore, considering the scenario of multiple clients, you are manipulating WiFi through shell command while the auto-connect program is running in the background. The confliction between the two clients will be inevitable if there is no WiFi state machine. Therefore I develop the new WiFi manager which can handle the problems mentioned above.

**Comment by kelvincheung** [comment]
> Any plans for supporting the enterprise security modes?



> Any plans for supporting the enterprise security modes?

Well, it depends on the user's requirement.
Through WPA Enterprise is important for WiFi, I'm not sure whether it is necessary for IoT devices.
Besides, lots of EAP methods need to be implemented due to IEEE 802.1x standard. This will consume much more memoy and let the WiFi manager become hostap perhaps.

**Comment by dleach02** [comment]
I'm not fully aware of all of the capabilities of these offloaded modules but it seems like they manage most of connection. I like what you are presenting here because it starts to enable into Zephyr more of the WiFi functionality that would be needed for platforms where there is WiFi IP but it isn't an offload like we currently have. Eventually, I would like to see Zephyr have something similar to what Linux has with an upper/lower MAC abstraction that would also need the layer you are presenting here.

My comment for normalizing the interfaces is mainly a desire to have a common configuration/management interface for WiFi that supports various levels of off-loading.

**Comment by GAnthony** [comment]
There is some interesting functionality added here, and glad to see the introduction of a WiFi control API.

Along the lines of the comments by @jukkar and @dleach02, I'm trying to see how existing NET_OFFLOAD or SOCKET_OFFLOAD WiFi drivers would work with this.

This PR introduces a separate subsystem, and so (naturally) duplicates some of the existing functionality found around current WiFi drivers based on `struct net_wifi_mgmt_offload`:
* WiFi driver apis:   scan, connect, disconnect, ap_enable/ap_disable APIs and NET_EVENT_WIFI_ commands already defined in `include/net/wifi_mgmt.h` - can that be leveraged/extended?
* WiFi Shell:   subsys/net/l2/wifi/wifi_shell.c: - can this be extended as well?
* WiFi API:   wifi_register_<*>_notifier() - can this leverage existing net_mgmt subsystem?
* WiFi API:   wifi_sta_<*>/wifi_ap_<*> apis - If a NET_OFFLOAD driver is configured in the system, could these APIs be made to vector to those drivers - it would be nice to have a public API for them too.

**Comment by GAnthony** [comment]
Also, on SoftAP:  Is the plan for this to support WiFi provisioning in the future? 
That could be interesting, as currently each vendor has their own method (though some vendor's methods may be more secure, if the passwords are negotiated via the offload chip and stored in secure flash).

**Comment by kelvincheung** [comment]
> There is some interesting functionality added here, and glad to see the introduction of a WiFi control API.
> 
> Along the lines of the comments by @jukkar and @dleach02, I'm trying to see how existing NET_OFFLOAD or SOCKET_OFFLOAD WiFi drivers would work with this.
> 
> This PR introduces a separate subsystem, and so (naturally) duplicates some of the existing functionality found around current WiFi drivers based on `struct net_wifi_mgmt_offload`:
> 
> * WiFi driver apis:   scan, connect, disconnect, ap_enable/ap_disable APIs and NET_EVENT_WIFI_ commands already defined in `include/net/wifi_mgmt.h` - can that be leveraged/extended?

As previously mentioned, I've ever tried to adapt our WiFi device for WiFi management before. However, those APIs aren't blocked APIs, and expose both WiFi commands (scan, connect, disconnect) and corresponding events to the users instead. The users have to handle synchronization by themselves. Moreover, it can't support multiple clients without synchronous mechanism and state machine. I can't see any possibility to leveraged/extended.

> * WiFi Shell:   subsys/net/l2/wifi/wifi_shell.c: - can this be extended as well?
> * WiFi API:   wifi_register_<*>_notifier() - can this leverage existing net_mgmt subsystem?

The whole WiFi management is just a bundle of net_mgmt hooks, and has no control logic at all. In contrast, the intention of introducing the WiFI manager is to provide a general purpose and dependent subsystem for WiFi control plane with full function including blocked APIs and status machine. I think the question is what kind of WiFi management does Zephyr need.

> * WiFi API:   wifi_sta_<_>/wifi_ap_<_> apis - If a NET_OFFLOAD driver is configured in the system, could these APIs be made to vector to those drivers - it would be nice to have a public API for them too.

The WiFi manager only cares about the control plane and has nothing to do with the data plane. Hence it can absolutely support NET_OFFLOAD drivers.

**Comment by kelvincheung** [comment]
> if the passwords are negotiated via the offload chip and stored in secure flash).

Yes, the WiFi manager will be able to handle l2 packets, even IEEE 802.1x in the near future if the memory footprint is not a problem.

File: samples/wifi/connect/README.rst [hunk 3]
```diff
@@ -0,0 +1,27 @@
+.. _wifi_sta_connect:
+
+Wi-Fi: Auto Connect
+####################
+
+Overview
+********
+
+Application demonstrating usage of STA connect APIs of Wi-Fi manager.
```
Comment (dbkinder) [inline samples/wifi/connect/README.rst]: Use a complete sentence, and fix some reST formatting (need a blank line before a list):
```
This sample demonstrates using the Wi-Fi manager STA connect APIs
to connect a board to an access point (AP), following these steps:

1. Get STA config to obtain the specified AP
2. Open STA
3. Scan the specified AP
4. Connect to the specified AP
```

File: samples/wifi/connect/README.rst [hunk 4]
```diff
@@ -0,0 +1,27 @@
...
+
+* An access point with IEEE 802.11 a/b/g/n/ac support
+
+Building and Running
+********************
+
+This sample can be found under :zephyr_file:`samples/wifi/connect` in
+the Zephyr tree.
+
+See :ref:`wifi samples section <wifi-samples>` for details.
```
Comment (dbkinder) [inline samples/wifi/connect/README.rst]: There are no build or run details in this reference document?

File: samples/wifi/repeater/README.rst [hunk 5]
```diff
@@ -0,0 +1,31 @@
+.. _wifi_repeater:
+
+Wi-Fi: Repeater
+####################
+
+Overview
+********
+
+Application demonstrating usage of STA&AP APIs of Wi-Fi manager.
```
Comment (dbkinder) [inline samples/wifi/repeater/README.rst]:
```
This sample demonstrates using the Wi-Fi manager STA and AP APIs
to start a Wi-Fi repeater, following these steps:

1. Get STA config to obtain the specified AP
2. Open STA
3. Scan the specified AP
4. Connect to the specified AP
5. Get AP config
6. Open AP
7. Start AP

Note that the packet forwarding is accomplished by Wi-Fi firmware.
```

File: samples/wifi/rtt/README.rst [hunk 6]
```diff
@@ -0,0 +1,29 @@
+.. _wifi_sta_rtt:
+
+Wi-Fi: RTT range
+####################
+
+Overview
+********
+
+Application demonstrating usage of STA RTT range of Wi-Fi manager.
```
Comment (dbkinder) [inline samples/wifi/rtt/README.rst]:
```
This sample demonstrates using the Wi-Fi manager STA
Round-Trip-Time (RTT) range APIs to measure the distance
to nearby Wi-Fi routers, following these steps:

1. Get STA config
2. Open STA
3. Scan the APs which support RTT
4. Assemble RTT range request by scan results
5. Send the RTT range request
6. Receive the RTT range response
```

File: samples/wifi/wifi.rst [hunk 7]
```diff
@@ -0,0 +1,10 @@
+.. _wifi-samples:
```
Comment (dbkinder) [inline samples/wifi/wifi.rst]: You reference this doc as having more build and run details, but there are none here.

File: samples/wifi/connect/README.rst (hunk 4 above)
Comment (dbkinder) [inline samples/wifi/connect/README.rst]: change to ``Wi-Fi samples section``

File: samples/wifi/repeater/README.rst [hunk 8]
```diff
@@ -0,0 +1,31 @@
...
+
+* An access point with IEEE 802.11 a/b/g/n/ac support
+
+Building and Running
+********************
+
+This sample can be found under :zephyr_file:`samples/wifi/repeater` in
+the Zephyr tree.
+
+See :ref:`wifi samples section <wifi-samples>` for details.
```
Comment (dbkinder) [inline samples/wifi/repeater/README.rst]: Change to ``Wi-Fi samples section``

File: samples/wifi/rtt/README.rst [hunk 9]
```diff
@@ -0,0 +1,29 @@
...
+
+* An access point with IEEE 802.11 a/b/g/n/ac support
+
+Building and Running
+********************
+
+This sample can be found under :zephyr_file:`samples/wifi/rtt` in
+the Zephyr tree.
+
+See :ref:`wifi samples section <wifi-samples>` for details.
```
Comment (dbkinder) [inline samples/wifi/rtt/README.rst]: Change to ``Wi-Fi samples section``

**Review by dbkinder — CHANGES_REQUESTED** [review]

File: samples/wifi/rtt/README.rst (hunk 9 above)
Comment (kelvincheung) [reply samples/wifi/rtt/README.rst]: @dbkinder 
I've reworked the sample documents.
Could you please review this part again?
Thanks!

File: samples/wifi/connect/README.rst [hunk 10]
```diff
@@ -0,0 +1,76 @@
...
+in this example we will build it for the ``96b_ivy5661`` board:
+
+.. zephyr-app-commands::
+   :zephyr-app: samples/wifi/connect
+   :board: 96b_ivy5661
+   :goals: build flash
+   :compact:
+
+To run this sample, the STA must be configured in advance.
+To specified an open AP, please follow the command below.
```
Comment (dbkinder) [inline samples/wifi/connect/README.rst]:
```suggestion
To specify an open AP, use the following command:
```

File: samples/wifi/connect/README.rst [hunk 11]
```diff
@@ -0,0 +1,76 @@
...
+
+.. code-block:: console
+
+    wifimgr set_config sta -n "MERCURY_3BC4" -p "" -a 10000
+
+To specified an WPA/WPA2-PSK secured AP, please follow the command below.
```
Comment (dbkinder) [inline samples/wifi/connect/README.rst]:
```suggestion
To specify a WPA/WPA2-PSK secured AP, use the following command:
```

File: samples/wifi/repeater/README.rst [hunk 12]
```diff
@@ -0,0 +1,105 @@
...
+   :zephyr-app: samples/wifi/connect
+   :board: 96b_ivy5661
+   :goals: build flash
+   :compact:
+
+To run this sample, both STA and AP must be configured in advance.
+
+Step 1: Config STA
+==================
+To specified an open AP, please follow the command below.
```
Comment (dbkinder) [inline samples/wifi/repeater/README.rst]:
```suggestion
To specify an open AP, use the following command:
```

File: samples/wifi/repeater/README.rst [hunk 13]
```diff
@@ -0,0 +1,105 @@
...
+
+.. code-block:: console
+
+    wifimgr set_config sta -n "MERCURY_3BC4" -p "" -a 10000
+
+To specified an WPA/WPA2-PSK secured AP, please follow the command below.
```
Comment (dbkinder) [inline samples/wifi/repeater/README.rst]:
```suggestion
To specify a WPA/WPA2-PSK secured AP, use the following command:
```

**Review by dbkinder — CHANGES_REQUESTED** [review]
looking better...

**Comment by nashif** [comment]
stale, closing. Reopen when there are updates...

**Pull request closed without merging**